ceph/ceph_admin/helper.py

```diff
@@ -391,6 +393,117 @@ def generate_rgw_spec(self, spec):
+        Return:
+            out, err    String values
+
+        Raises:
+            CommandFailed   when there is an execution failure
+        """
+        out, err = node.exec_command(**kwargs)
+
+        out = out.read().decode() if isinstance(out, ChannelFile) else out
+        err = err.read().decode() if isinstance(err, ChannelFile) else err
+        return out, err
+
+    def generate_snmp_spec(self, spec):
+        """
+        Return spec content for snmp-destination
+
+        Args:
+            spec (Dict): snmp-destination service spec config
+
+        Returns:
+            service_spec (Str)
+
+        Example::
+
+        specs:
+              - service_type: snmp-destination
+                spec:
+                  credentials:
+                    snmp_v3_auth_username: myadmin
+                    snmp_v3_auth_password: mypassword
+                    engine_id: fsid
+        """
+        template = self._get_template("snmp")
+        destination_node = spec["spec"].pop("snmp_destination", None)
+        node = get_node_by_id(self.cluster, destination_node)
+        if destination_node:
+            spec["spec"]["snmp_destination"] = self.get_addr(node) + ":162"
+        node_installer = get_node_by_id(self.cluster, "node1")
+        cmd = "cephadm shell ceph fsid"
+        out, err = self.exec_command(
+            node_installer,
+            sudo=True,
+            cmd=cmd,
+        )
+        LOG.info(f"fsid: {out}")
+        engine_id = out.replace("-", "")
+        if engine_id:
+            spec["spec"]["engine_id"] = engine_id
+        return template.render(spec=spec)
+
+    def generate_snmp_destination_conf(self, spec):
+        """
+        Return spec content for snmp-gateway service
+
+        Args:
+            spec (Dict): snmp-gateway service spec config
+
+        Returns:
+            service_spec (Str)
+
+        Example::
+
+            spec:
+                - service_type: snmp-gateway
+                  service_name: snmp-gateway
+                  placement:
+                    count: 1
+                  spec:
+                    credentials:
+                      snmp_v3_auth_username: <user_name>
+                      snmp_v3_auth_password: <password>
+                    engine_id: <fsid>
+                    port: 9464
+                    snmp_destination: node
+                    snmp_version: V3
+
+        """
+        template = self._get_template("snmp_destination")
+        node_installer = get_node_by_id(self.cluster, "node1")
+        cmd = "cephadm shell ceph fsid"
+        out, err = self.exec_command(
+            node_installer,
+            sudo=True,
+            cmd=cmd,
+        )
+        LOG.info(f"fsid: {out}")
+        fsid = out.replace("-", "")
+        engine_id = fsid[0:32]
+        if engine_id:
+            spec["spec"]["engine_id"] = engine_id
+        LOG.info(f"fsid:{engine_id}")
+        spec["spec"]["snmp_v3_auth_username"] = spec["spec"]["credentials"].pop(
```

Comment:
i think we could avoid these two assigning, if we pass `snmp_v3_auth_username` and `snmp_v3_auth_password` directly and nothing to process here.